### PR TITLE
MBQL Refactor: Combine various Field clauses into one new clause

### PR DESCRIPTION
**Pull request by camsaul**
Part of the #14057 wish list. This change has been on my wish list for years now, but I have an actual reason to do it -- implementing #9934 is super difficult without being able to attach arbitrary info for Field clauses.

### Background

The original version of MBQL (known as the "Structured Query" language at the time, which was a little confusing) just referred to all fields by integer ID. e.g.

```clj
{:filter ["STARTS_WITH" 1 "abc"]}
```

That made clauses like this ambiguous:

```clj
{:filter ["=" 1 2]} ; is 2 a Field, or the number 2?
```

To clear up ambiguity and to let you filter Fields against Fields, we added the `:field-id` clause to make it explicit that you were referring to a Field, not an Integer:

```clj
{:filter [:= [:field-id 1] [:field-id 2]]} ; Field 1 == Field 2
```

We soon added a couple of new types of Field clauses, `:fk->` and `:datetime-field`, both of which wrap the original `:field-id`:

```clj
;; refer to Field 2 from a different Table, use Field 1 from the current Table to perform the join
[:fk-> [:field-id 1] [:field-id 2]]

;; bucket Field 3 by month
[:datetime-field [:field 3] :month]
```

So far things were still pretty reasonable, the worst you'd have to do is something like

```clj
[:datetime-field [:fk-> [:field-id 1] [:field-id 2]] :month]
```

Things started to get out-of-hand IMO when we continued to add more Field clause types that just wrapped everything else. We added `:binning-strategy`:

```clj
[:binning-strategy <field> :num-buckets 10]
```

then `:field-literal`, to refer to a Field from a nested native source query:

```clj
[:field-literal "my_field" :type/Text]
```

then we added `:joined-field` to specify the Table you're _explicitly_ joining against that is the source of a Field:

```clj
[:joined-field "my_join" [:field-id 4]]
```

This ends up getting really hairy when you combine things. This is an actual real clause you can use right now:

```clj
[:binning-strategy
 [:datetime-field
  [:joined-field "my_join"
   [:field-literal "my_field" :type/DateTimeWithLocalTZ]]
  :month]
 :num-bins 10]
```

And even with all of those clauses, we still can't pass around arbitrary extra information in a way that would make it easy to implement performance optimizations. If we ever try to add *another* new clause, the whole house of cards is going to come crashing down.

### The proposal

Combine *all* of the Field clauses into a single new `:field` clauses with the schema

```clj
[:field id-or-name options-map]
```

Here are some before & after examples:

```clj
[:field-id 1] 
=> 
[:field 1 nil]

[:field-literal "my_field" :type/Text] 
=> 
[:field "my_field" {:base-type :type/Text}]

[:datetime-field [:field 1] :month] 
=> 
[:field 1 {:temporal-unit :month}]

[:binning-strategy [:field 1] :num-bins 10] 
=>
[:field 1 {:binning {:strategy :num-bins, :num-bins 10}}]

[:joined-field "my_join" [:field 1]] 
=> 
[:field 1 {:join-alias "my_join}]

[:fk-> [:field 1] [:field 2]] 
=> 
[:field 2 {:source-field 1}]

[:binning-strategy
 [:datetime-field
  [:joined-field "my_join"
   [:field-literal "my_field" :type/DateTimeWithLocalTZ]]
  :month]
 :num-bins 10]
=>
[:field "my_field" {:base-type :type/DateTimeWithLocalTZ, :join-alias "my_join", :binning {:strategy :num-bins, :num-bins 10}}]
```

### CODE REVIEW

**Comment by camsaul** [comment]
It's easy enough to "upgrade" the old-style clauses to the new clause on the backend -- I have this all working. I still need to update a lot of QP middleware to handle the new clause and update a lot tests as well -- there are about 400 backend tests left to update (which would take another day or so)

**Comment by camsaul** [comment]
Backend stuff is ready for review

**Comment by dpsutton** [comment]
i guess the work is already done now but i wonder if the diff could have been simpler to identify `[:field 1]` with `[:field 1 nil]`. Seems this might be a common usecase and is a straightfoward bit of sugar. That said, The diff is far too large to really look at anything. The one thing I'd worry about, when they were more s-expression and could wrap a single thing,  it might have been pretty easy to figure out the "order" of things, ie, which wraps which, and which are _allowed_ to wrap each other. Going to this flat map style, is that gone? Might someone make a map with stuff in it that's not valid even though its easily expressible? Or do we have to worry about an instance of `[:X [:Y [:field-id 1]]` being different some `[:Y [:X [:field-id 1]]` and the use of a map no longer lets us detect that (`[:field-id 1 {:X true :Y true}]` or some such)?

**Comment by camsaul** [comment]
@dpsutton all the clauses had to be wrapped in a certain order. So you would have had to remember to do `[:X [:Y [:field-id 1]]` as opposed to `[:Y [:X [:field-id 1]]]`, or you'd get an error. So by combining moving the `:X` and `:Y` stuff that previously wrapped the original clause into an options map, you can't do things the wrong way anymore. You can't make a Field clause that's satisfies the schema that's invalid, unless you do something like give it a `:join-alias` for a join that doesn't exist. It's much easier to construct correct MBQL with the new clause and much harder to construct incorrect MBQL, since you don't have to know the clause ordering rules

**Comment by camsaul** [comment]
@dpsutton I *would* have supported `[:field 1]` as the same as `[:field 1 nil]`, but the MBQL pattern matching utils that is used in almost every single QP namespace would consider those to be different clauses. The resulting QP code is a lot simpler if there's only one arity of the clause to deal with

**Review by dpsutton — APPROVED** [review]

**Comment by camsaul** [comment]
@ariya @paulrosenzweig could you review the frontend changes in this PR? It's almost all ready except for a few more tests to fix

**Review by paulrosenzweig — APPROVED** [review]
👏 👏 👏  This is really awesome. It was a massive undertaking to clean things up in one go. Bravo!

---

It's likely I missed them given the size of the PR, but if they're not in there already, we should have tests around migration. Some specific cases:
- Column settings
- Column settings on a dashcard
- Column settings on columns from nested questions
- Pivot table partitions
- Table column hiding
- Conditional formatting
- Dashboard parameter mapping
- Native question field filters

If there's no normalization on save via the API, I'd say we should have these be Cypress tests that set an "old style" question via API and checks that it still renders correctly.

**Comment by paulrosenzweig** [comment]
The only part of the new style I don't like is that we have to type sniff to distinguish fields from field literals. That should work but feels error prone. Did you consider flagging field literals in the options instead?

File: frontend/test/metabase/modes/components/drill/UnderlyingRecordsDrill.unit.spec.js [hunk 1]
```diff
@@ -95,14 +91,14 @@ describe("UnderlyingRecordsDrill", () => {
...
-        ["joined-field", "User", ["field-id", PEOPLE.ID.id]],
+        ["field", ORDERS.USER_ID.id, null],
+        ["field", PEOPLE.ID.id, { "join-alias": "User" }],
       ],
     };
     const query = ORDERS.query()
       .join(join)
       .aggregate(["count"])
-      .breakout(["joined-field", "User", ["field-id", PEOPLE.STATE.id]]);
+      .breakout(["field", PEOPLE.STATE.id, { "source-field": "User" }]);
```
[the remark this answers is not in the document]
Comment (camsaul) [reply frontend/test/metabase/modes/components/drill/UnderlyingRecordsDrill.unit.spec.js:101]: Thanks. Not sure why this test didn't fail

File: frontend/src/metabase-lib/lib/Dimension.js [hunk 2]
```diff
@@ -236,22 +242,12 @@ export default class Dimension {
...
 
-  foreign(dimension: Dimension): FKDimension {
-    return new FKDimension(
-      this,
-      [dimension.mbql()],
-      this._metadata,
-      this._query,
-    );
+  foreign(dimension: Dimension): FieldDimension {
+    return null;
```
[the remark this answers is not in the document]
Comment (camsaul) [reply frontend/src/metabase-lib/lib/Dimension.js:244]: I don't think they are. I copied them over from the old `DateTimeFieldDimension` and `FKDimension` classes. I'll remove them and see what breaks

**Comment by camsaul** [comment]
@paulrosenzweig RE sniffing stuff: there's two new methods in the new `FieldDimension` class, `isIntegerFieldId()` and `isFieldStringName()` that handle all the sniffing for everyone else... I'll have to try to make sure everything is using that instead of doing things themselves. I'm working on abstracting out the MBQL representation from all the other code so nobody else has to know about it. e.g. if we wanted to revert all the MBQL changes tomorrow we'd only have to change `FieldDimension` and none of the other non-test code

File: frontend/src/metabase-lib/lib/Dimension.js [hunk 3]
```diff
@@ -408,366 +407,544 @@ export default class Dimension {
...
+   */
+  static normalizeOptions(options: {}): {} {
+    if (!options) {
+      return null;
     }
-  }
+    // remove null values from the map.
+    Object.keys(options).forEach(k => {
+      if (options[k] === null || options[k] === undefined) {
+        delete options[k];
```
[the remark this answers is not in the document]
Comment (camsaul) [reply frontend/src/metabase-lib/lib/Dimension.js:464]: thanks, I wasn't sure what the best way to do this was. I've been having to Stack Overflow a lot of stuff that I could do in my sleep in Clojure

**Comment by camsaul** [comment]
@paulrosenzweig I added normalization coming out of the DB for all Card & DashboardCard `visualization_settings`, and it's covering everything I have locally correctly. It walks the whole map and normalized every old-style clause it finds. `:parameters` and related columns are already normalized on the way out of the DB. I *think* everything should be covered, but we can deploy to stats and make sure nothing is broken... part of the reason I wanted to get this refactor in at the beginning of the release cycle

File: frontend/src/metabase-lib/lib/Dimension.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (camsaul) [reply frontend/src/metabase-lib/lib/Dimension.js:464]: I think I can use `_.omit()` here http://underscorejs.org/#omit

File: frontend/src/metabase-lib/lib/Dimension.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (camsaul) [reply frontend/src/metabase-lib/lib/Dimension.js:244]: ok, so `foreign` is still used, but not `datetime`. Removed `datetime`

File: frontend/src/metabase-lib/lib/Dimension.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulrosenzweig) [reply frontend/src/metabase-lib/lib/Dimension.js:464]: Oh good call! I hand't realized `_.omit` allowed a predicate.

**Comment by paulrosenzweig** [comment]
> It walks the whole map and normalized every old-style clause it finds.

The JSON-serialized field refs might trip this up. Column formatting and pivot table partitions are two places where that happens.

It's probably worth having tests around some of those cases specifically. I expect some saved questions out in the wild will never be updated and rely on this migration continuing to work for years.

File: test/metabase/models/card_test.clj [hunk 4]
```diff
@@ -210,3 +210,58 @@
...
+    (testing "Don't normalize key-value pairs in maps that could be interpreted as MBQL clauses"
+      (let [original {:field-id 1}]
+        (f original original)))))
+
+(deftest normalize-visualization-settings-test
+  (test-visualization-settings-normalization
+   (fn [original expected]
+     (mt/with-temp Card [card {:visualization_settings original}]
+       (is (= expected
+              (db/select-one-field :visualization_settings Card :id (u/the-id card))))))))
```
Comment (camsaul) [inline test/metabase/models/card_test.clj:267]: @paulrosenzweig I added tests for normalization of serialized Card/DashboardCard field references as keys in viz settings here (specifically the `json/generate-string` part at the top). Are the serialized keys used anywhere else besides `visualization_settings.column_settings`?

**Comment by camsaul** [comment]
@paulrosenzweig this is green no so I'm going to go ahead and merge it in so we don't end up in merge-conflict hell since the PR is so big. Once it's up on stats I'll try to go test things and make sure everything is working properly and add more normalization test cases/fixes as needed

File: test/metabase/models/card_test.clj (hunk 4 above)
Comment (paulrosenzweig) [reply test/metabase/models/card_test.clj:267]: > Are the serialized keys used anywhere else besides visualization_settings.column_settings?

Sorry slow responding to that. I had thought both conditional formatting and pivot tables did, but I just checked and neither do! So thankfully I guess it's just column settings.

**Pull request merged**